### ISSUE TITLE
Audio Playback on local device stops working after call disconnects

### ISSUE THREAD

**Issue by dpraymond** [twilio/twilio-video-ios#167]
We have an application that requires us to play some sound locally during a call.  We are currently using an AVPlayer object to playback the sound on the local device and we are seeing a couple of issues.

When not in a call playing sounds seems to be quite loud and coming from the loudspeaker at whatever hardware volume setting we have set on the iPad.  Once a call is connected, that local audio volume drops about 30% or so even though the physical volume  on the iPad stays the same.  The other issue we are seeing is that once the call is disconnected and the user drops back into the waiting area, all local audio stops working completely, no audio playback can be heard.  If we reconnect the call, the Microphone audio works and we can communicate with the remote client, but our local sound playback generates no sound at all anymore.  If we restart the app we can get local sound working again.

We use a single AVAudioEngine for the entire application, currently the local sound is played back using an AVAudioPlayerNode that is connected to the AVAudioEngine mainMixerNode.  There is no error generated that we can see in this case, just that local audio playback seems to be non functional.

On the Simulator we don't seem to experience the same problem, local audio continues to work after a call disconnects, but on actual hardware this problem is 100% reproducible.

**Comment by piyushtank** [comment]
@dpraymond Thanks for reaching out. 

When VoIP call is running, iOS prioritizes the VoIP audio and suppresses the volume of audio played by AVPlayer. The recommended way is to use mix both audio and play it out. We have AVAudioEngineExample audio device to demonstrate this use case.

We have a work-in-progress PR that demonstrates how to play music even after the call is disconnected - https://github.com/twilio/video-quickstart-ios/pull/589 Try it out and let us know if you have any questions.

**Comment by dpraymond** [comment]
Thanks for the quick reply, I'm currently working to integrate the AVAUdioEngineExample into our repo, we are using Swift and I've had a few issues with compilation errors around the Bridging header but I'll power through as it seems like this is the only way to get this working.

Do you know if there is any way, other than using .measurement mode to suppress the audio processing on the device?  There is a use case where we want to disable all hardware audio processing on the iPad, we were using .measurement mode for this but from what I've been able to discern from reading up on it, this will cause all audio to route to the receiver instead of the loudspeaker despite using PortOverride.

**Comment by piyushtank** [comment]
@dpraymond If you want to suppress the hardware audio processing, instead of VoiceProcessingIO, you can try using RemoteIO with `AVAudioSessionCategoryPlayAndRecord` category and `AVAudioSessionCategoryOptionDefaultToSpeaker` with `measurement` mode. Personally, I haven't tried it, but it's worth giving a shot, I think. 

I want to highlight that not using VoiceProcessingIO in a VoIP call may cause echo issues in your app.

**Comment by dpraymond** [comment]
Okay, I believe I saw another thread about that issue, I'll give it a try.

We will make sure people know they should use headphones when in that mode to limit Echo as much as possible.  I've managed to integrate the ExampleAVAUdioEngineDevice and I integrated your patches for continuousMusic.  We aren't playing music per se, but we have a few things that generate sound, one of these is a keyboard that plays notes and another element is a metronome.  I was able to get the keyboard to work, but it only seems to work when the call is connected.  After implmenting your patch for continuous music the keyboard doesn't work when run locally and I actually get a sigabrt at 

`` assert(mixedAudioBufferList != NULL);`` in ExampleAVAudioEngineDeviceRecordCallback.

To make this work for my app I added in function that gives my application access to the audioEngine so we can register our sampler against it, does this make sense?  How can I make this work when there is no call connected.

Thanks.

**Comment by piyushtank** [comment]
@dpraymond Can you confirm you are calling `playMusic` with `continuous` argument set to `true`? 

`mixedAudioBufferList` is created [here](https://github.com/twilio/video-quickstart-ios/blob/master/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L596).

**Comment by dpraymond** [comment]
@piyushtank To simplify, since I will always require the ability to play local audio when not in a call I hard-coded continuousMusic to be true all the time

```
- (id)init {
    self = [super init];

    if (self) {
        /*
         * Initialize rendering and capturing context. The deviceContext will be be filled in when startRendering or
         * startCapturing gets called.
         */

        // Initialize the rendering context
        self.renderingContext = malloc(sizeof(AudioRendererContext));
        memset(self.renderingContext, 0, sizeof(AudioRendererContext));

        // Setup the AVAudioEngine along with the rendering context
        if (![self setupPlayoutAudioEngine]) {
            NSLog(@"Failed to setup AVAudioEngine");
        }

        // Initialize the capturing context
        self.capturingContext = malloc(sizeof(AudioCapturerContext));
        memset(self.capturingContext, 0, sizeof(AudioCapturerContext));
        self.capturingContext->bufferList = &_captureBufferList;

        // Setup the AVAudioEngine along with the rendering context
        if (![self setupRecordAudioEngine]) {
            NSLog(@"Failed to setup AVAudioEngine");
        }

        [self setupAVAudioSession];
        self.continuousMusic = YES;
    }

    return self;
}
```

**Comment by piyushtank** [comment]
@dpraymond Can you change it to lazy instantiation like [this](https://github.com/twilio/video-quickstart-ios/blob/dc9454289a5dab35e5b1342ec9ec67573f2c205b/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L439-L449) ? 
For the first time when you play music, in the patch, we setup and start the AVAudioEngine nodes and then set `continuousMusic` to true. Subsequent music play request, the patch doesn't set up the AVAudioEngine nodes, see [this](https://github.com/twilio/video-quickstart-ios/blob/dc9454289a5dab35e5b1342ec9ec67573f2c205b/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L569) and [this](https://github.com/twilio/video-quickstart-ios/blob/dc9454289a5dab35e5b1342ec9ec67573f2c205b/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L677). If you set `self.continuousMusic = YES;` in `init`, the AVAudioEngine player node to get started, hence no audio played.

**Comment by dpraymond** [comment]
@piyushtank Thanks so much for the input, I really appreciate it.

I have a final question, since i'm moving the AvAudioEngine into the ExampleAvAudioEngineDevice we were previously using code such as
```
// if we have a tap already remove it - Prevents a crash
        inputNode.removeTap(onBus: bus)
        inputNode.installTap(onBus: bus, bufferSize: bufferSize, format: format) { buffer, time in
            guard let averageLevel = self.averageLevel else { return }

```
to tap into the microphone input for some sound analysis for a tuner application, this no longer works even though I'm using the AudioEngine from the ExampleAVAudioEngineDevice, the tap get's installed successfully but we never receive any buffer day.  Is there some way you can think to do this with the way the Mic and Output audio are being mixed in ExampleAvAudioEngineDevice?

Thanks!

**Comment by dpraymond** [comment]
> @dpraymond Can you change it to lazy instantiation like [this](https://github.com/twilio/video-quickstart-ios/blob/dc9454289a5dab35e5b1342ec9ec67573f2c205b/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L439-L449) ?
> For the first time when you play music, in the patch, we setup and start the AVAudioEngine nodes and then set `continuousMusic` to true. Subsequent music play request, the patch doesn't set up the AVAudioEngine nodes, see [this](https://github.com/twilio/video-quickstart-ios/blob/dc9454289a5dab35e5b1342ec9ec67573f2c205b/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L569) and [this](https://github.com/twilio/video-quickstart-ios/blob/dc9454289a5dab35e5b1342ec9ec67573f2c205b/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m#L677). If you set `self.continuousMusic = YES;` in `init`, the AVAudioEngine player node to get started, hence no audio played.

@piyushtank I've implemented the above using your latest commit to ExampleAVAudioEngineDevice and I'm seeing the playMusic call being activated but I still don't hear anything locally while not in a call, not the example music or any of my local sounds, once a call has connected and then disconnected I can continue to play sounds, but on startup there is no sound until connected to a room.

Here is my Initialization code

```
var audioDevice =  ExampleAVAudioEngineDevice()
    private var file: AVAudioFile! = nil
    private var buffer: AVAudioPCMBuffer! = nil
    
    init(blob: Blob) {
        self.handleIncomignURL = program.receivedSharedVideo
        self.blob= blob
       
        TwilioVideoSDK.audioDevice = audioDevice
        if let audioEngine = audioDevice.getAudioEngine() {
            audioEngine.stop()
            audioEngine.attach(audioPlayerNode)
            audioEngine.attach(reverb)
            audioEngine.attach(echo)
            audioEngine.attach(audioMixer)
            audioEngine.attach(micMixer)
            audioEngine.attach(sampler)
                   
            audioEngine.connect(audioMixer, to: audioEngine.mainMixerNode, format: nil)
            audioEngine.connect(sampler, to: audioEngine.mainMixerNode, format: nil)
            
            let url = Bundle.main.resourceURL!.appendingPathComponent("metronome.wav")
            file = try! AVAudioFile(forReading: url)
            buffer = AVAudioPCMBuffer(
                pcmFormat: file.processingFormat,
                frameCapacity: AVAudioFrameCount(file.length)
            )
           
            
            try! file.read(into: buffer)
            
             audioEngine.connect(
                 audioPlayerNode, to: audioEngine.mainMixerNode, format: file.processingFormat
             )
         
            try! audioEngine.start()
            audioPlayerNode.play()
        }
        super.init()
        loadSF2PresetIntoSampler(0)
        audioDevice.playMusic(true)
```

**Comment by piyushtank** [comment]
@dpraymond The [PR](https://github.com/twilio/video-quickstart-ios/pull/589) successfully plays audio before connecting a call also. Can you prepare a branch on top of it and share it, I can take a look at the audio device implementation to see where the problem is?

**Comment by dpraymond** [comment]
> @dpraymond The [PR](https://github.com/twilio/video-quickstart-ios/pull/589) successfully plays audio before connecting a call also. Can you prepare a branch on top of it and share it, I can take a look at the audio device implementation to see where the problem is?

@piyushtank What  device implementation are you looking for?  I can't share our whole project unfortunately as the source code belongs to a customer and isn't opening sourcing it.  I'm happy to share any of the Twilio initalization code though, I pasted the assignment of the audioDevice code above and how I plugged in our sampler into the existing AudioEngine ( which seems to work once a call has been established ).  I only seem to hear the music playback though after a call starts.  

```
func prepareLocalMedia() {
        
        // default to voice chat mode on initial connect
        let captureSession = AVCaptureSession()
        captureSession.automaticallyConfiguresApplicationAudioSession = false
        setAudioMode()
        
        // We will share local audio and video when we connect to the Room.
        
        // Create an audio track.
        if localAudioTrack == nil {
            DispatchQueue.main.async {
                self.localAudioTrack = LocalAudioTrack(options: nil, enabled: true, name: "Microphone")
                self.localAudioAvailable = true
                
                if self.localAudioTrack == nil {
                    self.logMessage(messageText: "Failed to create audio track")
                    self.localAudioAvailable = false
                }
            }
        }
        
        // Create a video track which captures from the camera.
        if localVideoTrack == nil {
            DispatchQueue.main.async {
                self.previewView = VideoView(frame: CGRect.zero, delegate: self)
                self.startPreview()
                self.localVideoAvailable = true
            }
        }
        
    }
```

And our connect  setup and execution

```
func setupTwilioRoom() {
        // setup DataTrack
        let dataTrackOptions = DataTrackOptions { (builder) in
            builder.isOrdered = true
            builder.name = "DataTrack"
        }
        
        self.localDataTrack = LocalDataTrack(options: dataTrackOptions)
        
        if let accessTokenConst = accessToken {
            connectOptions = ConnectOptions(token: accessTokenConst) { (builder) in
                // Use the local media that we prepared earlier.
                builder.audioTracks = self.localAudioTrack != nil ? [self.localAudioTrack!] : [LocalAudioTrack]()
                builder.videoTracks = self.localVideoTrack != nil ? [self.localVideoTrack!] : [LocalVideoTrack]()
                
                // Use the preferred audio codec
                if let preferredAudioCodec = OpusCodec() {
                    builder.preferredAudioCodecs = [preferredAudioCodec]
                }
                
                // Use the preferred video codec
                if let preferredVideoCodec = Vp8Codec() {
                    builder.preferredVideoCodecs = [preferredVideoCodec]
                }
                
                /*
                 // Use the preferred encoding parameters
                 if let encodingParameters = Settings.shared.getEncodingParameters() {
                 builder.encodingParameters = encodingParameters
                 }
                 */
                builder.encodingParameters = EncodingParameters(audioBitrate: 16, videoBitrate: 0)
                // The name of the Room where the Client will attempt to connect to. Please note that if you pass an empty
                // Room `name`, the Client will create one for you. You can get the name or sid from any connected Room.
                
                builder.region = "gll"
                
                // The name of the Room where the Client will attempt to connect to. Please note that if you pass an empty
                // Room `name`, the Client will create one for you. You can get the name or sid from any connected Room.
                builder.roomName = self.roomCode
                self.localDataTrack = LocalDataTrack(options: dataTrackOptions)
                builder.dataTracks = [self.localDataTrack]
            }
        }
        else {
            errorMsg = "Unable to contact service, please try again or contact support if this continues."
        }
        
    }
    
    func connectToRoom()
    {
        logMessage(messageText: "Connecting to room")
      
        if let filledConnectOptions = connectOptions {
            self.room = TwilioVideoSDK.connect(options: filledConnectOptions, delegate: self)
        }
    }

```

**Comment by piyushtank** [comment]
@dpraymond Given that our sample code is able to achieve the functionality, I suspect the bug is in your app's AudioDevice implementation code. Without looking at the AudioDevice implementation code, it's hard to find the problem area. Can you create a branch top of the PR to replicate the problem?

**Comment by dpraymond** [comment]
> @dpraymond Given that our sample code is able to achieve the functionality, I suspect the bug is in your app's AudioDevice implementation code. Without looking at the AudioDevice implementation code, it's hard to find the problem area. Can you create a branch top of the PR to replicate the problem?

@piyushtank 
I'm simply using the defualt ExampleAVAudioEngineDevice from the branch you posted, I removed the two functions I added to expose the audioEngine's and for now am just going with the stock implementation.   There are no differences between the code in my project and the branch 

```
git branch
* (HEAD detached at origin/task/ISDK-3191-music)
* 
 git diff project-swiftui/Project/AudioDevices/ExampleAVAudioEngineDevice.m ./demo_twilio/AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```

**Comment by piyushtank** [comment]
@dpraymond Instead of just calling `[self playMusic:YES]` from `ExampleAVAudioEngineDevice`'s `init`, can you call it asynchronously on main queue - 
```
- (id)init {
    self = [super init];

    if (self) {
        /*
         * Initialize rendering and capturing context. The deviceContext will be be filled in when startRendering or
         * startCapturing gets called.
         */

        // Initialize the rendering context
        self.renderingContext = malloc(sizeof(AudioRendererContext));
        memset(self.renderingContext, 0, sizeof(AudioRendererContext));

        // Setup the AVAudioEngine along with the rendering context
        if (![self setupPlayoutAudioEngine]) {
            NSLog(@"Failed to setup AVAudioEngine");
        }

        // Initialize the capturing context
        self.capturingContext = malloc(sizeof(AudioCapturerContext));
        memset(self.capturingContext, 0, sizeof(AudioCapturerContext));
        self.capturingContext->bufferList = &_captureBufferList;
        
        // Setup the AVAudioEngine along with the rendering context
        if (![self setupRecordAudioEngine]) {
            NSLog(@"Failed to setup AVAudioEngine");
        }
        
        [self setupAVAudioSession];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self playMusic:YES];
        });
    }

    return self;
}
```

**Comment by dpraymond** [comment]
@piyushtank 
I'm seeing the following error in the logs

2021-02-05 14:35:28.805276-0500 [1420:447456]  ca_require: SanityCheck(newDesc) InvalidFormat
2021-02-05 14:35:28.805316-0500 [1420:447456] Could not set stream format on input bus!
2021-02-05 14:35:28.807096-0500 [1420:447456]  ca_require: SanityCheck(newDesc) InvalidFormat
2021-02-05 14:35:28.807133-0500 [1420:447456] Could not set stream format on input bus!

Any idea why these allocations would be failing?

**Comment by piyushtank** [comment]
@dpraymond Sounds like your app sets invalid audio format. There is a lot of back and forth on this thread, please create a branch on the audio device to replicate these errors and share. We won't be able to help debug without knowing your implementation of audio device.

**Comment by dpraymond** [comment]
> @dpraymond Sounds like your app sets invalid audio format. There is a lot of back and forth on this thread, please create a branch on the audio device to replicate these errors and share. We won't be able to help debug without knowing your implementation of audio device.

@piyushtank What code do you need to see?  Just the ExampleAVAudioEngienDevice.m and .h?

**Comment by piyushtank** [comment]
@dpraymond Yes. ExampleAVAudioEngienDevice.m and .h If you like, you can send me at my email ptank@twilio.com

**Comment by dpraymond** [comment]
@piyushtank 
Thank you for the support, I've sent an email with the relevant code attached including our Twilio handling code.

**Comment by piyushtank** [comment]
@dpraymond It was great talking to you over the live call. Closing the ticket as the issue has been resolved. 
Feel free to reach out to us if you have any questions.

**Comment by Sephaq** [comment]
Hello I understand this conclusion was reached however not denoted here, I'm currently facing the same issue, Our app s using v4.6.0 of the TwilioVideo.
When we open the LocalAudioTrack we lose access to sound in other parts of the app, when simply playing a sound file.
I've tried adding the AVAudioEngineExample but it did not help and I'm currently lost.
I've tested this by disabling the LocalAudioTrack and then the sound file plays normally.
